File: python/2454.py
from typing import List

class Solution:
    def secondGreaterElement(self, nums: List[int]) -> List[int]:
        s1 = []
        s2 = []
        ans = [-1] * len(nums)
        for i, num in enumerate(nums):
            while s2 and num > s2[-1][0]:
                ans[s2[-1][1]] = num
                s2 = s2[:-1]
            x = 0
            while s1 and num > s1[-1][0]:
                x += 1
            s2.extend(s1[-x:])
            s1 = s1[:-x]
            s1.append([num, i])
        return ans

File: go/2454.go
package main

func secondGreaterElement(nums []int) []int {
	s1 := []int{}
	s2 := []int{}
	ans := make([]int, len(nums))
	for i := range ans {
		ans[i] = -1
	}
	for i, num := range nums {
		for len(s2) > 0 && num > nums[s2[len(s2)-1]] {
			ans[s2[len(s2)-1]] = num
			s2 = s2[:len(s2)-1]
		}
		x := len(s1) - 1
		for x >= 0 && num > nums[s1[x]] {
			x--
		}
		s2 = append(s2, s1[x+1:]...)
		s1 = append(s1[:x+1], i)
	}
	return ans
}
